<a href="https://colab.research.google.com/github/bbellaborges/Desafio-CD-DIT/blob/main/Desafio_DIT_Sec_de_Sa%C3%BAde_RJ.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ✨Exploração e Análide de Dados

In [1]:
from google.colab import files #carregando a base de dados
uploaded = files.upload()

Saving Dados_ficha.csv to Dados_ficha.csv


## Carregar os dados e visualizar uma amostra
Aqui, vamos importar a biblioteca pandas, a partir da qual faremos uso de funções para analisar nossos dados. Carregaremos os dados e armazenaremos na variável df, por fim, veremos as primeiras linhas de cada uma das colunas da tabela para termos uma **ideia geral de como os dados estão estruturados**.

Isso vai nos permitir verificar se os dados foram carregados de forma correta, além de entender sua organização e as colunas que temos.

In [2]:
import pandas as pd #importando biblioteca para explorar e analisar os dados

df = pd.read_csv("Dados_ficha.csv") #carrega dados

#Exibe amostra dos dados
pd.set_option('display.max_columns', None) #para exibir todas as colunas
print(df.head()) #Visualizar os primeiros 5 casos de cada variável

                            id_paciente    sexo  obito           bairro  \
0  cd6daa6e-038d-4952-af29-579e62e07f97    male      0  Parada de Lucas   
1  ad6cecb2-3a44-49ab-b2f3-6f9ffc2e2ec7    male      0      Cidade Nova   
2  54e834e7-e722-4daa-8909-cf917a1247e6    male      0       Santa cruz   
3  c6a71e5e-0933-48d1-9d5a-8f448dc37f71  female  False            Bangu   
4  d5262a3c-e5d3-4195-a46b-0acd2533e1d6    male      0       Santíssimo   

  raca_cor                          ocupacao      religiao luz_eletrica  \
0   Branca                     Não se aplica  Sem religião         True   
1    Parda         Assistente Administrativo    Evangélica            1   
2    Parda                     Não se aplica      Católica            1   
3   Branca  Representante Comercial Autônomo  Sem religião            1   
4   Branca               Técnico Eletricista         Outra         True   

             data_cadastro            escolaridade nacionalidade  \
0  2021-06-24 00:00:00.000  Fu

O carregamento do dataset foi bem-sucedido, confirmando que temos 100.000 registros e 31 colunas. Isso já nos dá uma noção do volume de dados que estamos lidando e da necessidade de um processamento eficiente.

## Classificação dos Campos: tipos das variáveis
Vamos exibir o tipo de dado contido em cada coluna/variável, isto é, se é um int, float, object ou datetime. Com isso, saberemos se cada variável tem o tipo correto (de acordo com o esperado para sua descrição, por exemplo, pra idade se espera um número), assim podemos evitar problemas na análise. No caso de colunas numéricas tratadas como object(texto), por exemplo, saberemos que temos um problema na **ingestão dos dados**.


In [8]:
print("\nTipos das variáveis:\n", df.dtypes) #Identifica os tipos de cada coluna


Tipos das variáveis:
 id_paciente                               object
sexo                                      object
obito                                     object
bairro                                    object
raca_cor                                  object
ocupacao                                  object
religiao                                  object
luz_eletrica                              object
data_cadastro                             object
escolaridade                              object
nacionalidade                             object
renda_familiar                            object
data_nascimento                           object
em_situacao_de_rua                        object
frequenta_escola                           int64
meios_transporte                          object
doencas_condicoes                         object
identidade_genero                         object
meios_comunicacao                         object
orientacao_sexual                         obje

A classificação das variáveis mostrou a diversidade de informações coletadas. Temos variáveis numéricas, categóricas e booleanas, cada uma com características específicas que precisarão de tratamentos distintos. A distinção correta dessas variáveis facilitará análises mais direcionadas na Parte 2, onde avaliaremos possíveis inconsistências.

## Identificação de valores ausentes

Vamos buscar quantos valores ausentes (NaN) existem em cada coluna. Faremos isso porque dados faltantes podem impactar nossas análises, além disso, colunas com um grande número de dados ausentes podem **indicar problemas na coleta**.

In [9]:
print("\nValores ausentes:\n", df.isnull().sum()) #Identifica valores ausentes e a completude do dataset


Valores ausentes:
 id_paciente                                0
sexo                                       0
obito                                      0
bairro                                     0
raca_cor                                   0
ocupacao                                   0
religiao                                   0
luz_eletrica                               0
data_cadastro                              0
escolaridade                               0
nacionalidade                              0
renda_familiar                             0
data_nascimento                            0
em_situacao_de_rua                         0
frequenta_escola                           0
meios_transporte                           0
doencas_condicoes                          0
identidade_genero                        952
meios_comunicacao                          0
orientacao_sexual                          0
possui_plano_saude                         0
em_caso_doenca_procura             

A identificação de colunas com valores ausentes revelou que algumas variáveis possuem taxas significativas de missing values. Esse dado é fundamental para avaliar a completude da informação e entender se a ausência dos dados pode impactar análises posteriores. Questões como a forma de coleta e a obrigatoriedade do preenchimento podem ser exploradas na próxima etapa.

## Classificação das variáveis

O código abaixo permite classificar as variáveis em grupos, a fim de facilitar nossa análise. Como?
Separando as colunas em grupos de acordo com seu significado e o tipo do dado que contém:
*   Identificadores - códigos, não usados para análises estatísticas;
*   Categóricas - contém categorias (ocupação, raça, sexo);
*   Binárias - Sim/Não, 0/1 ou Verdadeiro/Falso;
*   Numéricas - valores contínuos (altura, peso, pressão) ou discretos (idade);
*   Datas - informações temporais;

Para analisar cada tipo de variável separadamente, esse agrupamento é interessante.

In [10]:
# Identificadores
id_cols = ['id_paciente', 'updated_at', 'data_cadastro', 'data_atualizacao_cadastro']

# Categóricas
cat_cols = ['sexo', 'bairro', 'raca_cor', 'ocupacao', 'religiao', 'nacionalidade', 'escolaridade', 'renda_familiar', 'meios_transporte', 'doencas_condicoes',
            'identidade_genero', 'meios_comunicacao', 'orientacao_sexual', 'situacao_profissional', 'em_caso_doenca_procura', 'tipo']
# Binárias
bin_cols = ['obito', 'luz_eletrica', 'em_situacao_de_rua', 'frequenta_escola', 'possui_plano_saude',
            'familia_beneficiaria_auxilio_brasil', 'crianca_matriculada_creche_pre_escola', 'vulnerabilidade_social']

# Numéricas
num_cols = ['altura', 'peso', 'pressao_sistolica', 'pressao_diastolica', 'n_atendimentos_atencao_primaria', 'n_atendimentos_hospital']

# Datas
date_cols = ['data_nascimento']

## Explorando variáveis

### Variáveis categóricas

Para cada variável (coluna) categórica, vamos contar quantas vezes cada valor aparece. Isso vai nos permitir entender a distribuição dos dados, além de permitir identificar categorias díspares ou erros, tais como:
 Sexo = "Fem", "Feminino", "Female", "F". Isto é, escrita inconsistente.
 Religião com valores inesperados ou escritos de formas diferentes.

In [34]:
print("\nValores nas variáveis categóricas:")
for col in cat_cols:
  print(f"\n{col}:")
  print(df[col].value_counts(dropna=False))


Valores nas variáveis categóricas:

sexo:
sexo
female    55488
male      44512
Name: count, dtype: int64

bairro:
bairro
Campo Grande                     5873
Santa cruz                       4297
Bangu                            3963
Realengo                         3382
Guaratiba                        2856
                                 ... 
Vila Pacheco da Rocha               1
Parque São Pedro de Alcantara       1
Manassu                             1
Sapê                                1
Jardim Passa Vinte                  1
Name: count, Length: 748, dtype: int64

raca_cor:
raca_cor
Parda       50856
Branca      35113
Preta       11373
Amarela      2619
Indígena       36
Não             3
Name: count, dtype: int64

ocupacao:
ocupacao
Não se aplica                                                      47000
Vendedor de Comércio Varejista                                      3091
Empregado Doméstico Diarista                                        2328
Administrador               

Algumas categorias apresentam valores inesperados, como "Não" na variável de raça/cor ou categorias estranhas na religião.
Algumas variáveis possuem um número excessivo de categorias, como bairro (748 categorias) e ocupação (1.355 categorias), o que pode indicar problemas de padronização (ou não, a depender, por exemplo, do território em questão).
Algumas categorias contêm caracteres especiais e diferentes formatos de escrita, o que pode impactar o tratamento dos dados.
Esses pontos serão investigados na Parte 2 para garantir a confiabilidade da informação.

### Variáveis Binárias

Vamos listar os valores de cada variável binárias para verificarmos se hpa problemas de formatação como, por exemplo, termos "0/1" misturado com "Sim/Não"

In [21]:
print("\nValores nas variáveis binárias:")
for col in bin_cols:
  print(f"\n{col}:")
  print(df[col].unique()) #Podemos ter "1", "True", "Sim" e etc.


Valores nas variáveis binárias:

obito:
['0' 'False' '1' 'True']

luz_eletrica:
['True' '1' '0' 'False']

em_situacao_de_rua:
['0' 'False' '1']

frequenta_escola:
[0 1]

possui_plano_saude:
['False' '1' '0' 'True']

familia_beneficiaria_auxilio_brasil:
['0' '1' 'False' 'True']

crianca_matriculada_creche_pre_escola:
['0' '1' 'False' 'True']

vulnerabilidade_social:
['0' '1' 'False' 'True']


As variáveis booleanas apresentam inconsistências nos formatos, como valores misturados entre "0/1", "True/False" e até strings como "Sim". Essa falta de padronização pode gerar problemas em análises futuras, exigindo um pré-processamento adequado.

### Variáveis Numéricas

Vamos calcular estatísticas das variáveis numéricas, tais como: média, valor mínimo, máximo e desvio padrão. Isso vai nos permitir veridicar se temos valores extremos (outliers) ou incosistentes com os dados. Se, por exemplo, a altura média estiver muito alta ou baixa, isso pode indicar erro nos dados.

In [23]:
print("\nEstatísticas das variáveis numéricas:")
print(df[num_cols].describe())


Estatísticas das variáveis numéricas:
             altura          peso  pressao_sistolica  pressao_diastolica  \
count  99975.000000  99816.000000       99960.000000        99983.000000   
mean     142.693589     63.801629         129.544068           79.378207   
std       38.659345     35.594173          21.406429           21.062140   
min        0.000000      0.000000           0.000000            0.000000   
25%      145.000000     50.700000         120.000000           70.000000   
50%      158.000000     67.900000         130.000000           80.000000   
75%      165.000000     82.000000         140.000000           87.000000   
max      810.000000    998.000000         900.000000          921.000000   

       n_atendimentos_atencao_primaria  n_atendimentos_hospital  
count                    100000.000000            100000.000000  
mean                          5.542710                 7.312930  
std                           5.204464                 9.801987  
min         

Os valores mínimos de altura, peso e pressão arterial são 0, o que pode indicar erros na coleta ou no armazenamento.
Os valores máximos de altura (810 cm) e peso (998 kg) são irreais, sugerindo erros de digitação ou problemas no sistema de registro.
A distribuição dos atendimentos hospitalares e na atenção primária tem uma grande variação, o que pode indicar diferentes perfis de uso dos serviços de saúde.
Esses achados serão explorados na Parte 2 para entender a origem desses problemas e como corrigi-los.

## Identificar dados incoerentes com as variáveis

Aqui, verificamos possíveis erros nos nossos dados.
Vamos tomar as variáveis: altura, peso e idade.
Vamos filtrar pacientes com altura ou peso fora do esperado, lembrando sempre de observar se temos um erro de digitaçao contornável (ex: 1800 cm ao invés de 180 cm) ou não. Vamos também converter a data de nascimento em idade em busca de pessoas com idade negativa ou maior que 120 anos.

 O peso normal de um recém-nascido varia entre 2.500 e 4.500 gramas, com média de 3.400 gramas.
 A altura média de um bebê recém-nascido a termo é de 50 cm.

In [31]:
print("\nAlturas menores que 50 cm ou maiores que 250 cm:")
print(df[(df['altura'] < 50) | (df['altura'] >250)][['altura']])

print("\nPesos menores que 3.4 kg ou maiores que 200 kg:")
print(df[(df['peso'] < 2.02) | (df['peso'] >200)][['peso']])


Alturas menores que 50 cm ou maiores que 250 cm:
       altura
40       48.0
53       47.0
614       0.0
615       0.0
616       0.0
...       ...
99820    17.5
99840    37.9
99870    48.0
99966    34.7
99986    46.0

[1571 rows x 1 columns]

Pesos menores que 3.4 kg ou maiores que 200 kg:
         peso
2972   657.00
2974   414.00
2986   954.00
2988   849.00
4138     1.95
...       ...
95067    0.00
95292    0.00
96235  302.00
97168  488.00
98183  578.00

[151 rows x 1 columns]


Altura: Foram encontrados 1.571 registros com valores menores que 50 cm ou maiores que 250 cm, o que pode indicar a existência de erros na coleta ou digitação.
Peso: Foram identificados 151 registros abaixo de 3.4 kg ou acima de 200 kg, o que também aponta para problemas similares.

In [33]:
df['data_nascimento'] = pd.to_datetime(df['data_nascimento'], errors = 'coerce')
df['idade'] = (pd.to_datetime('today') - df['data_nascimento']).dt.days / 365.25

print("\nIdades menores que 0 anos ou maiores que 120 anos:")
print(df[(df['idade'] < 0) | (df['idade'] >120)][['idade']])


Idades menores que 0 anos ou maiores que 120 anos:
Empty DataFrame
Columns: [idade]
Index: []


Nenhum valor extremo foi encontrado, indicando que essa variável pode estar com dados mais consistentes.
Esses achados são importantes para definir estratégias de limpeza e imputação dos dados na Parte 2.